### PR TITLE
feat!: Update WorkflowInfo; Add upsertSearchAttributes

### LINKED ISSUE #314: [Feature Request] Add missing continuedAsNewExecutionRunId to WorkflowInfo interface

### LINKED ISSUE #357: [Feature Request] Expose history size to Workflow code
[As discussed on our community slack](https://temporalio.slack.com/archives/C01DKSMU94L/p1635615168067800), we'd like to expose the number of history events and the aggregated payloads size to via `WorkflowInfo` so Workflow authors can make smarter decisions for when to continue as new.

### LINKED ISSUE #445: Add parent workflow information in `WorkflowInfo`

### LINKED ISSUE #576: [Feature Request] Return string for Workflow status
### Is your feature request related to a problem? Please describe.

We leverage `handle.describe()` to see the status of a workflow. Today, we have to build a wrapper around the response to translate the gRPC code (a number) to a meaningful status (a string).

### Describe the solution you'd like

Return a string type to represent status in the `WorkflowExecutionDescription`.

More explicitly, this:
```typescript
export interface WorkflowExecutionDescription {
    ...
    status: temporal.api.enums.v1.WorkflowExecutionStatus;
    ...
}

enum WorkflowExecutionStatus {
    WORKFLOW_EXECUTION_STATUS_UNSPECIFIED = 0,
    WORKFLOW_EXECUTION_STATUS_RUNNING = 1,
    WORKFLOW_EXECUTION_STATUS_COMPLETED = 2,
    WORKFLOW_EXECUTION_STATUS_FAILED = 3,
    WORKFLOW_EXECUTION_STATUS_CANCELED = 4,
    WORKFLOW_EXECUTION_STATUS_TERMINATED = 5,
    WORKFLOW_EXECUTION_STATUS_CONTINUED_AS_NEW = 6,
    WORKFLOW_EXECUTION_STATUS_TIMED_OUT = 7
}
```

turns into this:

```typescript
export interface WorkflowExecutionDescription {
  ...
  status: WorkflowExecutionStatus
  ...
}

export type WorkflowExecutionStatus = 'Canceled' | 'Completed' | 'Continued_As_New' | 'Failed' | ...
```

### PR DESCRIPTION
**Pull request by lorensr**
Working on broken tests, and will add more for new `WorkflowInfo` fields.

- Added and changed `WorkflowInfo` fields
- Add `upsertSearchAttributes` Workflow function
- Make Search Attributes always arrays
- Accept Dates for Search Attributes
- Convert ISO strings back to Dates when providing Search Attributes to user

## Fixes

- Fixes #314
- Fixes #445 
- Closes #576 
- Closes #357

### CODE REVIEW

File: packages/common/src/converter/search-attribute-payload-converter.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { ValueError } from '@temporalio/internal-workflow-common';
+import { PayloadConverter } from './payload-converter';
+import { JsonPayloadConverter } from './json-payload-converter';
+import { Payload, str } from './types';
+
+const jsonConverter = new JsonPayloadConverter();
+
+/**
+ *
+ */
```
Comment (Sushisource) [inline packages/common/src/converter/search-attribute-payload-converter.ts:10]: Empty docstring

File: packages/internal-workflow-common/src/retry-policy.ts [hunk 2]
```diff
@@ -71,3 +71,19 @@ export function compileRetryPolicy(retryPolicy: RetryPolicy): temporal.api.commo
     nonRetryableErrorTypes: retryPolicy.nonRetryableErrorTypes,
   };
 }
+
+export function uncompileRetryPolicy(
```
Comment (Sushisource) [inline packages/internal-workflow-common/src/retry-policy.ts:75]: uncompile is a bit of a weird term here. If you wanna stay consistent with above, `decompile` is probably more common

Alternatively, `extract`

File: packages/workflow/src/workflow.ts [hunk 3]
```diff
@@ -1133,3 +1134,52 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
...
+ *   CustomIntField: [2, 3],
+ *   CustomKeywordField: ['durable code']
+ * });
+ * ```
+ *
+ * would result in the Workflow having these Search Attributes:
+ *
+ * ```ts
+ * {
+ *   CustomIntField: [2, 3],
```
Comment (Sushisource) [inline packages/workflow/src/workflow.ts:1159]: Shouldn't this be `[1,2,3]` ? Or would the user need to take, append, then replace? If so, worth showing that example too.

Comment (lorensr) [reply packages/workflow/src/workflow.ts:1159]: No, I'm trying to make it clear here that upsert overwrites. I'll change the numbers. In the sample, I show upserting based on current value

**Review by bergundy — COMMENTED** [review]
Looks like I was in the middle of a review a while ago and never submitted it, might be outdated by now so feel free to ignore. I'll review again.

File: packages/common/src/converter/search-attribute-payload-converter.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+export class SearchAttributePayloadConverter implements PayloadConverter {
+  public toPayload(value: unknown): Payload | undefined {
+    if (value instanceof Array && value.length > 0 && value[0] instanceof Date) {
+      value = (value as Date[]).map((date) => date.toISOString());
```
Comment (bergundy) [inline packages/common/src/converter/search-attribute-payload-converter.ts:14]: Not required, this is automatically done by `JSON.stringify`

File: packages/internal-workflow-common/src/interfaces.ts [hunk 5]
```diff
@@ -35,3 +35,10 @@ export interface QueryDefinition<_Ret, _Args extends any[] = []> {
 
 /** Get the "unwrapped" return type (without Promise) of the execute handler from Workflow type `W` */
 export type WorkflowResultType<W extends Workflow> = ReturnType<W> extends Promise<infer R> ? R : never;
+
+/**
+ * If another SDK creates a Search Attribute that's not an array, we wrap it in an array.
+ *
+ * Dates are serialized as ISO strings.
+ */
+export type SearchAttributeValue = string[] | number[] | boolean[] | Date[];
```
Comment (bergundy) [inline packages/internal-workflow-common/src/interfaces.ts:44]: Might as well make it single value and wrap it in array everywhere:
```suggestion
export type SearchAttributeValue = string | number | boolean | Date;
```

File: packages/internal-workflow-common/src/workflow-options.ts [hunk 6]
```diff
@@ -47,18 +47,18 @@ export interface BaseWorkflowOptions {
...
-   * Specifies additional indexed information in result of list workflow. The type of value should
-   * be a primitive (e.g. string, number, boolean). For dates, use Date.toISOString()
+   * Specifies additional indexed information to attach to the Workflow Execution. More info:
+   * https://docs.temporal.io/docs/typescript/search-attributes
    *
-   * Values are always converted using {@link defaultPayloadConverter}, even when a custom data converter is provided.
+   * Values are always converted using {@link JsonPayloadConverter}, even when a custom data converter is provided.
    */
-  searchAttributes?: Record<string, string | number | boolean>;
+  searchAttributes?: Record<string, SearchAttributeValue>;
```
Comment (bergundy) [inline packages/internal-workflow-common/src/workflow-options.ts:61]: If you take my suggestion above this would be:
```suggestion
  searchAttributes?: Record<string, SearchAttributeValue[]>;
```

File: packages/test/src/integration-tests.ts [hunk 7]
```diff
@@ -631,7 +693,7 @@ export function runIntegrationTests(codec?: PayloadCodec): void {
       searchAttributePayloadConverter.fromPayload(
         execution.raw.workflowExecutionInfo!.searchAttributes!.indexedFields!.CustomIntField!
       ),
-      3
+      [3]
```
Comment (bergundy) [inline packages/test/src/integration-tests.ts:696]: I'd also check the not "raw" value

File: packages/worker/src/worker.ts [hunk 8]
```diff
@@ -907,20 +919,75 @@ export class Worker {
...
+                          this.options.loadedDataConverter,
+                          lastCompletionResult?.payloads
+                        ),
+                        lastFailure: await decodeOptionalFailureToOptionalError(
+                          this.options.loadedDataConverter,
+                          continuedFailure
+                        ),
                         taskQueue: this.options.taskQueue,
-                        isReplaying: activation.isReplaying,
+                        more: {
```
Comment (bergundy) [inline packages/worker/src/worker.ts:974]: Why nest this?

File: packages/worker/src/worker.ts [hunk 9]
```diff
@@ -1028,7 +1095,7 @@ export class Worker {
             ifaceName,
             fnName,
           });
-        } else if (dep.callDuringReplay || !info.isReplaying) {
+        } else if (dep.callDuringReplay || !info.unsafe.isReplaying) {
```
Comment (bergundy) [inline packages/worker/src/worker.ts:1098]: It's late so I might be missing something but I'm pretty sure we never update `isReplaying`

Comment (bergundy) [reply packages/worker/src/worker.ts:1098]: Don't have to fix in this PR, could open an issue but this is pretty critical.

File: packages/workflow/src/interfaces.ts [hunk 10]
```diff
@@ -1,39 +1,164 @@
...
+): ParentWorkflowInfo | undefined {
+  if (!parent) {
+    return undefined;
+  }
+
+  return {
+    /* eslint-disable @typescript-eslint/no-non-null-assertion */
+    workflowId: parent.workflowId!,
+    runId: parent.runId!,
+    namespace: parent.namespace!,
```
Comment (bergundy) [inline packages/workflow/src/interfaces.ts:160]: Instead of non null assertion let's fail at runtime with an exception or our users will get undefined values if something changes.
In general we shouldn't use non null assertions in the codebase, only in tests.

File: packages/workflow/src/workflow.ts [hunk 11]
```diff
@@ -1133,3 +1134,52 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
...
+  }
+
+  const mergedSearchAttributes = { ...state.info.searchAttributes, ...searchAttributes };
+  if (!mergedSearchAttributes) {
+    throw new Error('searchAttributes must be a non-null Record<string, SearchAttributeValue>');
+  }
+
+  state.pushCommand({
+    upsertWorkflowSearchAttributesCommandAttributes: {
+      seq: state.nextSeqs.upsertSearchAttributes++,
```
Comment (bergundy) [inline packages/workflow/src/workflow.ts:1179]: Why is this required @Sushisource?

File: packages/workflow/src/internals.ts [hunk 12]
```diff
@@ -403,6 +405,7 @@ export class State {
     signalWorkflow: 1,
     cancelWorkflow: 1,
     condition: 1,
+    upsertSearchAttributes: 1,
```
Comment (bergundy) [inline packages/workflow/src/internals.ts:408]: What is this used for?

File: packages/workflow/src/workflow.ts [hunk 13]
```diff
@@ -1133,3 +1134,52 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
...
+    throw new IllegalStateError('`state.info` should be defined');
+  }
+
+  const mergedSearchAttributes = { ...state.info.searchAttributes, ...searchAttributes };
+  if (!mergedSearchAttributes) {
+    throw new Error('searchAttributes must be a non-null Record<string, SearchAttributeValue>');
+  }
+
+  state.pushCommand({
+    upsertWorkflowSearchAttributesCommandAttributes: {
```
Comment (bergundy) [inline packages/workflow/src/workflow.ts:1178]: @Sushisource this should be called `upsertSearchAttributes` IMHO, we definitely don't need the `CommandAttributes` suffix, this is the only place where we use this pattern.

File: packages/common/src/converter/search-attribute-payload-converter.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+import { JsonPayloadConverter } from './json-payload-converter';
+import { Payload, str } from './types';
+
+const jsonConverter = new JsonPayloadConverter();
+
+/**
+ *
+ */
+export class SearchAttributePayloadConverter implements PayloadConverter {
+  public toPayload(value: unknown): Payload | undefined {
```
Comment (bergundy) [inline packages/common/src/converter/search-attribute-payload-converter.ts:12]: We should verify only valid types are passed in here (only arrays with single type)

File: packages/workflow/src/workflow.ts [hunk 15]
```diff
@@ -1133,3 +1134,52 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
...
+      seq: state.nextSeqs.upsertSearchAttributes++,
+      searchAttributes: mapToPayloads(searchAttributePayloadConverter, mergedSearchAttributes),
+    },
+  });
+
+  state.info.searchAttributes = mergedSearchAttributes;
```
Comment (bergundy) [inline packages/workflow/src/workflow.ts:1184]: Let's test that this is indeed updated

File: packages/workflow/src/workflow.ts [hunk 16]
```diff
@@ -1133,3 +1134,52 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
...
+
+  const mergedSearchAttributes = { ...state.info.searchAttributes, ...searchAttributes };
+  if (!mergedSearchAttributes) {
+    throw new Error('searchAttributes must be a non-null Record<string, SearchAttributeValue>');
+  }
+
+  state.pushCommand({
+    upsertWorkflowSearchAttributesCommandAttributes: {
+      seq: state.nextSeqs.upsertSearchAttributes++,
+      searchAttributes: mapToPayloads(searchAttributePayloadConverter, mergedSearchAttributes),
```
Comment (bergundy) [inline packages/workflow/src/workflow.ts:1180]: Why do we need to pass in all the search attributes? Just pass in the new values and merge the maps in workflow memory.

File: packages/workflow/src/interfaces.ts [hunk 17]
```diff
@@ -1,39 +1,164 @@
...
+// If we get use cases / user requests for this,
+// expose startWorkflow.continueInitiator as WorkflowInfo.initiator: Initiator
+// export enum Initiator {
+//   Unspecified = 0,
+//   ContinueAsNew = 1,
+//   Retry = 2,
+//   Cron = 3,
+// }
+
+export function updateParentType(
```
Comment (bergundy) [inline packages/workflow/src/interfaces.ts:149]: Looks like this is only used in the worker, I wouldn't put this in the workflow package.

**Review by bergundy — COMMENTED** [review]
Overall nicely done.

Not sure we should rename `workflowType` to `type` at this point, it's a breaking change I'd rather avoid, users are relying on this. Also not sure we should have the `more` attribute, could just flatten the whole thing.

File: packages/workflow/src/workflow.ts (hunk 11 above)
Comment (Sushisource) [reply packages/workflow/src/workflow.ts:1179]: I suppose it isn't, really, but it's consistent and IMHO we should eventually be resolving this command so users can decide to await on it if they need to make sure the changes are persisted before proceeding with the workflow.

File: packages/workflow/src/workflow.ts (hunk 13 above)
Comment (Sushisource) [reply packages/workflow/src/workflow.ts:1178]: Sure, can do that.

**Comment by Sushisource** [comment]
> Overall nicely done.
> 
> Not sure we should rename `workflowType` to `type` at this point, it's a breaking change I'd rather avoid, users are relying on this. Also not sure we should have the `more` attribute, could just flatten the whole thing.

FWIW I was conflicted about `more`, thought about it, and decided it was pretty nice actually. Narrows down the amount of stuff you see when you autocomplete to just the things you're more likely to care about.

File: packages/worker/src/worker.ts (hunk 9 above)
Comment (lorensr) [reply packages/worker/src/worker.ts:1098]: We set it from the start workflow activation:

image

And then update it during future activations in `vm.ts` and `worker-interface.ts`:

image

image

File: packages/workflow/src/workflow.ts (hunk 16 above)
Comment (lorensr) [reply packages/workflow/src/workflow.ts:1180]: You mean we should just pass the new ones to core? @Sushisource does that work atm?

File: packages/workflow/src/workflow.ts (hunk 15 above)
Comment (lorensr) [reply packages/workflow/src/workflow.ts:1184]: Does this suffice? `test('upsertAndReadSearchAttributes'`

**Comment by lorensr** [comment]
I like `WorkflowInfo.type` since you know what type it is given the context, and `workflowInfo().workflowType` seems repetitive. `WorkflowInfo.more.namespace` is also breaking (used to be `WorkflowInfo.namespace`), but probably less used. I'm okay changing it back to `workflowType`.

I like `more` a lot for being able to differentiate in autocomplete and docs what the normal/common fields are, and reduce mental load when looking at the normal stuff. I'd also be interested in doing similar things elsewhere, like `WorkerOptions` is a lot / intimidating to take in, and takes time to search through to find the field you're looking for:

image

and many could be hidden behind `.throttling` or something.

File: packages/worker/src/worker.ts (hunk 9 above)
Comment (bergundy) [reply packages/worker/src/worker.ts:1098]: Yup which means it doesn't take affect in the main thread where the worker lives and it requires this info to run sink calls.

File: packages/workflow/src/workflow.ts (hunk 11 above)
Comment (bergundy) [reply packages/workflow/src/workflow.ts:1179]: That's not how search attributes work though, there's no way to await this command's completion and even if there was we'd need to change the language runtime to support that.

File: packages/workflow/src/workflow.ts (hunk 13 above)
Comment (bergundy) [reply packages/workflow/src/workflow.ts:1178]: https://github.com/temporalio/sdk-core/issues/332

File: packages/workflow/src/workflow.ts (hunk 15 above)
Comment (bergundy) [reply packages/workflow/src/workflow.ts:1184]: Yes, this would work.

File: packages/workflow/src/workflow.ts (hunk 16 above)
Comment (bergundy) [reply packages/workflow/src/workflow.ts:1180]: That should work, I'd just test and verify

Comment (bergundy) [reply packages/workflow/src/workflow.ts:1180]: It's the way it's supposed to work.

File: packages/workflow/src/workflow.ts [hunk 18]
```diff
@@ -1133,3 +1134,52 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
     throw new TypeError(`Invalid definition type: ${(def as any).type}`);
   }
 }
+
+/**
+ * Updates this Workflow's Search Attributes by merging the provided `searchAttributes` with the existing Search
```
Comment (bergundy) [inline packages/workflow/src/workflow.ts:1139]: Nice docstring

File: packages/test/src/workflows/return-search-attributes.ts [hunk 19]
```diff
@@ -0,0 +1,11 @@
+import { SearchAttributeValue, workflowInfo } from '@temporalio/workflow';
+
+export async function returnSearchAttributes(): Promise<Record<string, SearchAttributeValue[]> | undefined> {
+  const sa = workflowInfo().searchAttributes!; // eslint-disable-line @typescript-eslint/no-non-null-assertion
+  const datetime = (sa.CustomDatetimeField as Array<Date>)[0];
+  return {
+    ...sa,
+    datetimeType: [Object.getPrototypeOf(datetime).constructor.name],
+    datetimeInstanceofWorks: [datetime instanceof Date],
```
Comment (bergundy) [inline packages/test/src/workflows/return-search-attributes.ts:9]: Great test

File: packages/workflow/src/interfaces.ts [hunk 20]
```diff
@@ -1,41 +1,151 @@
...
+  };
+}
+
+export interface ParentWorkflowInfo {
+  workflowId: string;
+  runId: string;
+  namespace: string;
 }
 
+// If we get use cases / user requests for this,
```
Comment (bergundy) [inline packages/workflow/src/interfaces.ts:140]: Wanna track this in an issue?

File: packages/common/src/converter/search-attribute-payload-converter.ts (hunk 4 above)
Comment (bergundy) [reply packages/common/src/converter/search-attribute-payload-converter.ts]: Why did you resolve this?

File: packages/common/src/converter/search-attribute-payload-converter.ts [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+import { Payload, str } from './types';
+
+const jsonConverter = new JsonPayloadConverter();
+
+/**
+ * Converts Search Attribute values using JsonPayloadConverter
+ */
+export class SearchAttributePayloadConverter implements PayloadConverter {
+  public toPayload(values: unknown): Payload | undefined {
+    if (!(values instanceof Array)) {
```
Comment (bergundy) [inline packages/common/src/converter/search-attribute-payload-converter.ts:13]: I was going to ask why not use `Array.isArray` but there's no need in our case: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray#instanceof_vs_isarray

File: packages/common/src/converter/search-attribute-payload-converter.ts [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+    if (firstType === 'object') {
+      for (const value of values) {
+        if (!(value instanceof Date)) {
+          throw new ValueError(
+            `SearchAttribute values must arrays of strings, numbers, booleans, or Dates. This value ${value} is a ${typeof value}.`
+          );
+        }
+      }
+    } else {
+      for (const value of values) {
```
Comment (bergundy) [inline packages/common/src/converter/search-attribute-payload-converter.ts:28]: value could be undefined or bigint, we should should prevent that, let's have an explicit whitelist of types (string, number, boolean).

File: packages/worker/src/worker.ts (hunk 9 above)
Comment (lorensr) [reply packages/worker/src/worker.ts:1098]: https://github.com/temporalio/sdk-typescript/issues/667

File: packages/workflow/src/workflow.ts (hunk 11 above)
Comment (Sushisource) [reply packages/workflow/src/workflow.ts:1179]: I understand it's not how it works now, hence "should eventually". We can get rid of the seq until then, but, not exactly a big deal either way.

File: packages/workflow/src/interfaces.ts [hunk 23]
```diff
@@ -1,39 +1,140 @@
...
+
+    /**
+     * Milliseconds after which the Workflow Execution is automatically terminated by Temporal Server. Set via {@link WorkflowOptions.workflowExecutionTimeout}.
+     */
+    executionTimeout?: number;
+
+    /**
+     * Time at which the Workflow Execution expires
+     */
+    executionExpirationTime?: Date;
```
Comment (cretz) [inline packages/workflow/src/interfaces.ts:90]: Where does this value come from? Are we aware of any other SDK ever needing it?

File: packages/workflow/src/interfaces.ts [hunk 24]
```diff
@@ -1,39 +1,140 @@
...
+  };
+
+  /**
+   * These fields are non-deterministic. Workflow code should not do different things based on these fields.
+   */
+  unsafe: {
+    /**
+     * Whether a Workflow is replaying history or processing new events
+     */
+    isReplaying: boolean;
```
Comment (cretz) [inline packages/workflow/src/interfaces.ts:130]: I don't think this should be in workflow info. As a rule of thumb, I don't think workflow info should _ever_ change after the workflow is started from a user POV. I should be able to store the workflow info in a local variable and trust that it's immutable and doesn't change for the life of the workflow run. This property holds true in other SDKs I believe.

File: packages/workflow/src/interfaces.ts [hunk 25]
```diff
@@ -1,39 +1,140 @@
...
-   * Whether a Workflow is replaying history or processing new events
+   * Task queue this Workflow is executing on
    */
-  isReplaying: boolean;
+  taskQueue: string;
+
+  /**
+   * Less commonly used information
+   */
+  more: {
```
Comment (cretz) [inline packages/workflow/src/interfaces.ts:60]: Personally I think these can just move to the top level. There's no need to hide this info IMO. I can't think of a good reason to hide this stuff (a vertical scrollbar in intellisense is ok). Having said that, we shouldn't just add everything in workflow info, only valid stuff that represents information about the workflow run (_not_ the workflow task).

File: packages/workflow/src/interfaces.ts [hunk 26]
```diff
@@ -1,39 +1,140 @@
...
+
+    /**
+     * Cron Schedule for this Execution. Set via {@link WorkflowOptions.cronSchedule}.
+     */
+    cronSchedule?: string;
+
+    /**
+     * Milliseconds between Cron Runs
+     */
+    cronScheduleToScheduleInterval?: number;
```
Comment (cretz) [inline packages/workflow/src/interfaces.ts:120]: Has anyone asked for this value? Maybe we can leave it off.

File: packages/workflow/src/interfaces.ts (hunk 24 above)
Comment (bergundy) [reply packages/workflow/src/interfaces.ts:130]: You're right, we should move this from info, I originally missed this.

File: packages/workflow/src/workflow.ts (hunk 16 above)
Comment (bergundy) [reply packages/workflow/src/workflow.ts:1180]: Not sure what you mean by tested Core, you should test this against the server.

File: packages/client/src/workflow-client.ts [hunk 27]
```diff
@@ -809,7 +810,7 @@ export class WorkflowClient {
           searchAttributes: mapFromPayloads(
             searchAttributePayloadConverter,
             raw.workflowExecutionInfo!.searchAttributes?.indexedFields
```
Comment (bergundy) [inline packages/client/src/workflow-client.ts:812]: Doesn't have to be in this PR but let's get rid of all of these non null assertions

File: packages/common/src/converter/json-payload-converter.ts [hunk 28]
```diff
@@ -0,0 +1,37 @@
+import { ValueError } from '@temporalio/internal-workflow-common';
+import { PayloadConverterWithEncoding } from './payload-converters';
+import { encodingKeys, encodingTypes, METADATA_ENCODING_KEY, Payload, str, u8 } from './types';
+
+/**
+ * Converts between non-undefined values and serialized JSON Payload
```
[the remark this answers is not in the document]
Comment (bergundy) [reply packages/common/src/converter/json-payload-converter.ts:6]: It's not just non-undefined, it's JSON serializables, but yes, it explicitly ignores `undefined`.

File: packages/common/src/converter/search-attribute-payload-converter.ts [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * Converts Search Attribute values using JsonPayloadConverter
+ */
+export class SearchAttributePayloadConverter implements PayloadConverter {
+  public toPayload(values: unknown): Payload | undefined {
+    if (!(values instanceof Array)) {
+      throw new ValueError(`SearchAttribute value must be an array`);
+    }
+
+    const firstValue = values[0];
```
Comment (bergundy) [inline packages/common/src/converter/search-attribute-payload-converter.ts:18]: If you really want to be accurate with the error messages you could check that the array isn't empty or better yet, does empty array unset the value? Is unsetting supported?

**Review by bergundy — COMMENTED** [review]
Let's leave WorkflowInfo flat like the earth, revert the change from `workflowType` to `type` and in a followup PR move `isReplaying` out of `WorkflowInfo`.

File: packages/workflow/src/interfaces.ts (hunk 25 above)
Comment (lorensr) [reply packages/workflow/src/interfaces.ts:60]: > not the workflow task

Looks like `taskTimeout` is the only task-specific info? I like the idea of a Workflow being able to know what settings it was started with, whether it's just used for logging or impacts logic.

**Comment by bergundy** [comment]
You can add custom search attributes using the gRPC `operatorService` API.

We don't have a high level API for that yet.

There's also a need to work around the fact that the connection in the test environment isn't a concrete class (I'll fix that in a bit)

```ts
import { temporal } from '@temporalio/proto';
const { client } = testEnv;
(client.connection as Connection).operatorService.addSearchAttributes({
  namespace: client.options.namespace,
  searchAttributes: {
      MyAttribute: temporal.api.enums.v1.IndexedValueType.INDEXED_VALUE_TYPE_INT,
  },
});
```

**Comment by siawyoung** [comment]
> You can add custom search attributes using the gRPC `operatorService` API.
> 
> We don't have a high level API for that yet.
> 
> There's also a need to work around the fact that the connection in the test environment isn't a concrete class (I'll fix that in a bit)
> 
> ```ts
> import { temporal } from '@temporalio/proto';
> const { client } = testEnv;
> (client.connection as Connection).operatorService.addSearchAttributes({
>   namespace: client.options.namespace,
>   searchAttributes: {
>       MyAttribute: temporal.api.enums.v1.IndexedValueType.INDEXED_VALUE_TYPE_INT,
>   },
> });
> ```

Holy moly, that was a quick turnaround. And it worked too!

```
  await (testEnv.connection as Connection).operatorService.addSearchAttributes({
    searchAttributes: {
      Patch: iface.temporal.api.enums.v1.IndexedValueType.INDEXED_VALUE_TYPE_KEYWORD
    }
  })
```

(Also just a weird side note but why does `KEYWORD_LIST` still exist when everything is now treated as an array? I tried using `INDEXED_VALUE_TYPE_KEYWORD_LIST` at first but it threw an error with `protoEnumName should start with INDEXED_VALUE_TYPE_ prefix`)

**Comment by bergundy** [comment]
`KEYWORD_LIST` was added in server version 1.20, make sure you're using a new enough server.
We also plan to make all other search attributes single value only eventually.
